### PR TITLE
update cache_utilities for mac

### PR DESCRIPTION
**Pull request by buck54321**

### CODE REVIEW

File: client/webserver/site/cache_utilities.bash [hunk 1]
```diff
@@ -53,7 +59,11 @@ hashjssrc () { hashdir "${JS_DIR}" ; }
 # hashjssrc hashes the compiled js.
 hashjsdist () {
     cp "${JS_FILE}" js.tmp
-    sed -i 's/commitHash="[^"]*"//' js.tmp
+    if [[ "$OSTYPE" == "darwin"* ]]; then
+        sed -i '' 's/commitHash="[^"]*"//' js.tmp
+    else
+        sed -i 's/commitHash="[^"]*"//' js.tmp
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/webserver/site/cache_utilities.bash:65]: Oh yeah.

File: client/webserver/site/cache_utilities.bash [hunk 2]
```diff
@@ -5,14 +5,20 @@ CSS_FILE=${SITE_DIR}/dist/style.css
 JS_DIR=${SITE_DIR}/src/js
 JS_FILE=${SITE_DIR}/dist/entry.js
 
-hashfile () { sha1sum $1 | cut -d " " -f1 ; }
+if [[ "$OSTYPE" == "darwin"* ]]; then
+   hashfunc="shasum"
+else
+   hashfunc="sha1sum"
+fi
```
Comment (chappjc) [inline client/webserver/site/cache_utilities.bash:12]: Some of the decred folks seem to prefer using openssl as a portable approach.  `openssl sha1 -r blah.txt`
The `sha*sum` functions seem be be part of typical "coreutils" packags in NIX systems.  I've always felt like that's more common than having openssl installed personally, but it does make the app consistent so there's no question about syntax or output format.

File: client/webserver/site/cache_utilities.bash (hunk 1 above)
[the remark this answers is not in the document]
Comment (buck54321) [reply client/webserver/site/cache_utilities.bash:65]: I'm seeing some weird behavior with this recommendation implemented. `jsdistbuster` works correctly when I run it directly, but when run in `bust_caches.sh` it sets the wrong value.

File: client/webserver/site/cache_utilities.bash (hunk 2 above)
Comment (ukane-philemon) [reply client/webserver/site/cache_utilities.bash:12]: > I wonder is openssl available out of the box on Mac?

yh, but the latest macOS, `openssl version` shows Apple uses another flavour `LibreSSL`.

File: client/webserver/site/cache_utilities.bash [hunk 3]
```diff
@@ -52,10 +52,9 @@ hashjssrc () { hashdir "${JS_DIR}" ; }
 
 # hashjssrc hashes the compiled js.
 hashjsdist () {
-    cp "${JS_FILE}" js.tmp
-    sed -i 's/commitHash="[^"]*"//' js.tmp
-    HASH=$(hashfile js.tmp | cut -c1-8)
-    rm js.tmp
+    sed -i.tmp 's/commitHash="[^"]*"//' "${JS_FILE}"
+    HASH=$(hashfile "${JS_FILE}" | cut -c1-8)
+    rm "${JS_FILE}.tmp"
```
Comment (buck54321) [inline client/webserver/site/cache_utilities.bash:57]: Oh wait. This won't work. We **must** create a copy and discard it, and this edits the original in place. Doing `sed -i.tmp` doesn't create a new file to edit in place, it uses that path for a backup of the original.

Comment (chappjc) [reply client/webserver/site/cache_utilities.bash]: > Oh wait. This won't work. We **must** create a copy and discard it, and this edits the original in place. Doing `sed -i.tmp` doesn't create a new file to edit in place, it uses that path for a backup of the original.

Oh hah, got that backwards!

**Comment by chappjc** [comment]
LGTM, but I don't have access to a Mac currently.

**Comment by ukane-philemon** [comment]
> LGTM, but I don't have access to a Mac currently.

I can branch off this and test gh CI run.

**Comment by chappjc** [comment]
> > LGTM, but I don't have access to a Mac currently.
> 
> I can branch off this and test gh CI run.

I mean I think this PR is g2g if someone else with a Mac can just check that `./bust_caches.sh` runs on their machine.

**Comment by ukane-philemon** [comment]
> I mean I think this PR is g2g if someone else with a Mac can just check that ./bust_caches.sh runs on their machine.

Okay, testing.

**Review by martonp — APPROVED** [review]

**Comment by chappjc** [comment]
> But [fails CI ](https://github.com/ukane-philemon/dcrdex/actions/runs/5360380751/jobs/9725174655)run even after busting for this change [b0cbc74](https://github.com/decred/dcrdex/commit/b0cbc74d48da04b758cc6205ecbbdb1019abaf24)

So the buster you get don't work for CI?

**Comment by buck54321** [comment]
@ukaine tested same on a Mac and is passing. If you pull this last commit [aac3a2b](https://github.com/decred/dcrdex/pull/2410/commits/aac3a2b4aca57d63158d2cf9220184452c5665a5), go to the site directory, and run

```Bash
npm run build
source cache_utilities.bash
jsdistbuster
```

what do you get?

**Comment by ukane-philemon** [comment]
> So the buster values you get don't work for CI?

I guess so. I don't know if I was supposed to build before busting. I simply made changes to the js file, ran the bust script, and committed, then I pushed the changes to test CI but it failed.

**Comment by chappjc** [comment]
> > So the buster values you get don't work for CI?
> 
> I guess so. I don't know if I was supposed to build before busting. I simply made changes to the js file, ran the bust script, and committed, then I pushed the changes to test CI but it failed.

If you use bust_cashes.sh, it should build for you.

https://github.com/decred/dcrdex/blob/aac3a2b4aca57d63158d2cf9220184452c5665a5/client/webserver/site/bust_caches.sh

Which piece of the buster was wrong, first, second or both?

**Comment by chappjc** [comment]
Never forget to do `npm ci` before `npm run build` BTW.  That can change your output easily.  Never do `npm install`.

**Comment by ukane-philemon** [comment]
> If you use bust_cashes.sh, it should build for you.

Yeah, that was what I relied on.

> Never do npm install

Yh, I never do.

**Comment by chappjc** [comment]
> > If you use bust_cashes.sh, it should build for you.
> 
> Yeah, that was what I relied on.

Dang.  Try removing your node_modules folder manually I guess.

Also, I assume you're running either Node 18 or 20?

**Comment by chappjc** [comment]
> Which piece of the buster was wrong, first, second or both?

Also this @ukane-philemon.  If you run on the current commit here in buck's PR, which piece of the buster does it tweak?

File: client/webserver/site/src/html/bodybuilder.tmpl [hunk 4]
```diff
@@ -104,7 +104,7 @@
 {{end}}
 
 {{define "bottom"}}
-<script src="/js/entry.js?v=88986a77|0cdffb94"></script>
+<script src="/js/entry.js?v=ebcfcdea|a99362bc"></script>
```
Comment (chappjc) [inline client/webserver/site/src/html/bodybuilder.tmpl:107]: Getting the same values locally.

@ukane-philemon Do you have `core.autocrlf` set to something with git?

**Comment by ukane-philemon** [comment]
> Also, I assume you're running either Node 18 or 20?

I'm using node 20.3.1.

**Comment by ukane-philemon** [comment]
> > Which piece of the buster was wrong, first, second or both?
> 
> Also this @ukane-philemon. If you run on the current commit here in buck's PR, which piece of the buster does it tweak?

It seems the second part is the black sheep (`ebcfcdea|5e62b35d`) from https://github.com/decred/dcrdex/commit/b0cbc74d48da04b758cc6205ecbbdb1019abaf24

File: client/webserver/site/src/html/bodybuilder.tmpl (hunk 4 above)
Comment (ukane-philemon) [reply client/webserver/site/src/html/bodybuilder.tmpl:107]: > @ukane-philemon Do you have core.autocrlf set to something with git?

No, I don't.

**Comment by chappjc** [comment]
> It seems the second part is the black sheep (`ebcfcdea|5e62b35d`) from [b0cbc74](https://github.com/decred/dcrdex/commit/b0cbc74d48da04b758cc6205ecbbdb1019abaf24)

So at least we know the input part is fine, which means the `hashfile` and `hashdir` functions are OK for you.  So it could be an issue with `hashjsdist`.  Perhaps hack into `cache_utilities.bash` and add some temporary variables and echo statements, and comment the `rm` line so you can inspect and hash it manually after.  Do a diff of the js file to see if the sed worked as intended.

Anyway, you don't need to test CI with your own branch.  Just run bust_caches.sh on this branch until the values don't change from what's already committed.

**Comment by chappjc** [comment]
Another thing to try is to change "sed" to "perl -pe" to see if your sed is the culprit.

**Comment by ukane-philemon** [comment]
> ust run bust_caches.sh on this branch until the values don't change from what's already committed.

Alright.

**Comment by ukane-philemon** [comment]
> 

`perl -pe` worked.

**Review by ukane-philemon — APPROVED** [review]
G2G

**Review by chappjc — APPROVED** [review]

**Pull request merged**